rust/noosphere/src/ffi/petname.rs

```diff
@@ -1,47 +1,173 @@
 ) {
-    todo!();
+    error_out.try_or_initialize(|| {
+        noosphere.async_runtime().block_on(async {
+            sphere
+                .inner_mut()
+                .set_petname(petname.to_str(), did.map(|did| did.to_str().into()))
+                .await?;
+
+            Ok(())
+        })
+    });
 }
 
 #[ffi_export]
-/// Resolve a configured petname to a sphere version (a CID), via the DID that
-/// has been assigned to it. The returned value is a UTF-8, base64-encoded CIDv1
-/// string. If no DID has been assigned to the given petname, no value will be
-/// resolved.
+/// Resolve a configured petname, using the sphere idenitty - a DID - to
+/// determine a link - a CID - that is associated with it. The returned link is
+/// a UTF-8, base64-encoded CIDv1 string that may be used to resolve data from
+/// the IPFS content space. Note that this call will produce an error if no
+/// address has been assigned to the given petname.
 pub fn ns_sphere_petname_resolve(
-    noosphere: &NsNoosphereContext,
-    sphere_identity: char_p::Ref<'_>,
+    noosphere: &NsNoosphere,
+    sphere: &NsSphere,
     petname: char_p::Ref<'_>,
     error_out: Option<Out<'_, repr_c::Box<NsError>>>,
 ) -> Option<char_p::Box> {
-    todo!();
+    error_out.try_or_initialize(|| {
+        noosphere.async_runtime().block_on(async {
+            sphere
+                .inner()
+                .resolve_petname(petname.to_str())
+                .await?
+                .ok_or_else(|| anyhow!("No record resolved for petname '{}'", petname.to_str()))?
+                .to_string()
+                .try_into()
+                .map_err(|error: InvalidNulTerminator<String>| anyhow!(error).into())
+        })
+    })
+}
+
+#[ffi_export]
+/// Get an array of all of the petnames in a sphere at the current version.
+pub fn ns_sphere_petname_list(
+    noosphere: &NsNoosphere,
+    sphere: &NsSphere,
+    error_out: Option<Out<'_, repr_c::Box<NsError>>>,
+) -> c_slice::Box<char_p::Box> {
+    let possible_output = error_out.try_or_initialize(|| {
+        noosphere.async_runtime().block_on(async {
+            let petname_set = SphereWalker::from(sphere.inner()).list_petnames().await?;
+            let mut all_petnames: Vec<char_p::Box> = Vec::new();
+
+            for petname in petname_set.into_iter() {
+                all_petnames.push(
+                    petname
+                        .try_into()
+                        .map_err(|error: InvalidNulTerminator<String>| anyhow!(error))?,
+                );
+            }
+
+            Ok(all_petnames)
+        })
+    });
+
+    match possible_output {
+        Some(slugs) => slugs,
+        None => Vec::new(),
+    }
+    .into_boxed_slice()
+    .into()
+}
+
+#[ffi_export]
+/// Get an array of all of the petnames that changed in a given sphere since a
+/// given revision of that sphere (excluding the given revision). The revision
+/// should be provided as a UTF-8 base64-encoded CIDv1 string. If no revision is
```

Comment:
It's great, verbose, and necessary, to document the string types used on this interface, e.g. "UTF-8 base64-encoded CIDv1 string", wonder if we can consolidate that somehow in documentation without some extra opaque type in consuming platforms.

Reply:
We may be able to hoist some of this to module-level with sufficient work on the header transformer?

Reply:
Aye.. I'll file something more specific about "class" documentation that this could cover